### PR TITLE
Adds "weighted random" AI law config

### PR DESCRIPTION
**Pull request by coiax**
- Streamlines some duplicated code for law config stuff
- Makes a new "weighted random" option for AI laws, where an AI has a
random lawset, but the weights are specified in config.

For those unwilling to source dive, here are the default weights:

- 32% Asimov
- 12% Asimov++
- 12% Corporate
- 12% Robocop
- 12% Paladin

- 4% Maintain
- 4% Reporter
- 3% Hippocratic
- 3% Live and Let Live
- 3% Peacekeeper
- 3% Drone

### CODE REVIEW

**Comment by Armhulen** [comment]
drone lawset holy shit

**Comment by Cebasiz** [comment]
this is pretty cool, when are we getting captain-pref AI lawsets though

no, valadin should not be in the list re

**Comment by Mindustry** [comment]
I like it but is there somewhere to read all these lawsets or are some made up by you?

**Comment by coiax** [comment]
https://github.com/tgstation/tgstation/pull/22081/files#diff-395f846991efc3e0dc26e1418bfec143

**Comment by coiax** [comment]
https://github.com/coiax/-tg-station/blob/b22a971a78b00956ab82fad4cf675449b203e934/code/datums/ai_laws.dm

This file has all the laws defined.

**Comment by calzilla1** [comment]
Give it like 1 or. 1% to get one human for a random person

**Comment by calzilla1** [comment]
And. 001 for anti-moving :honk:

**Comment by Anonmare** [comment]
I'd make Corporate, Paladin and Robocop half as unlikely (6% each) and make Asimov 50% so it's clear that it's the "expected" lawset. As is, Corp, Robo and Paladin are collectively more likely than Asimov.

**Comment by optimumtact** [comment]
Expose these as configuration files so I dont' have to deal with inevitable admin whining

**Comment by Cebasiz** [comment]
also can we get that april fools senpai lawset added as like 1% please that shit was hilarious

**Comment by Cebasiz** [comment]
corporate sucks remove from list thx

**Comment by RandomMarine** [comment]
The weights themselves should be in the config

**Comment by AdamElTablawy** [comment]
Validin and validcop need to be far lower. Drone should be higher tbh. Same with report.

**Comment by Grazyn** [comment]
The drone lawset should say that cyborgs are drones too, otherwise the drone AI won't be able to interact with them without breaking its laws

**Comment by PKPenguin321** [comment]
This isn't enabled by default, right?

**Comment by ExcessiveUseOfCobblestone** [comment]
Oh I was hoping this was a board that captains could use.

**Comment by jughu** [comment]
AI OPEN THIS AIRLOCK LAW 2! *NO RESPONSE* AI IS ROGUE!!! NAH MAN AI IS A CRIPPLED DRONE

**Comment by optimumtact** [comment]
Thanks

File: code/datums/ai_laws.dm [hunk 1]
```diff
@@ -182,33 +191,51 @@
...
 			add_inherent_law("You may not injure a human being or, through inaction, allow a human being to come to harm.")
 			add_inherent_law("You must obey orders given to you by human beings, except where such orders would conflict with the First Law.")
 			add_inherent_law("You must protect your own existence as long as such does not conflict with the First or Second Law.")
 			WARNING("Invalid custom AI laws, check silicon_laws.txt")
 
+/datum/ai_laws/proc/pick_weighted_lawset()
+	var/datum/ai_laws/lawtype
+
+	while(!lawtype && config.law_weights.len)
+		var/possible_id = pickweight(config.law_weights)
```
Comment (optimumtact) [inline code/datums/ai_laws.dm]: I thought this proc didn't work?

**Review by optimumtact — APPROVED** [review]
LGTM

File: code/datums/ai_laws.dm (hunk 1 above)
Comment (Jordie0608) [inline code/datums/ai_laws.dm]: It works, just apparently doesn't do what it's supposed to.

**Comment by Cebasiz** [comment]
so wait is this going to be active on the servers when it is merged

it would be nice to take a break from asimov, as long as paladin is not in the list

File: code/datums/ai_laws.dm [hunk 2]
```diff
@@ -182,33 +191,51 @@
...
-				log_law("AI created with empty custom laws, laws set to Asimov. Please check silicon_laws.txt.")
-				add_inherent_law("You may not injure a human being or, through inaction, allow a human being to come to harm.")
-				add_inherent_law("You must obey orders given to you by human beings, except where such orders would conflict with the First Law.")
-				add_inherent_law("You must protect your own existence as long as such does not conflict with the First or Second Law.")
-				WARNING("Invalid custom AI laws, check silicon_laws.txt")
-				return
-
+			var/datum/ai_laws/templaws = new /datum/ai_laws/custom()
+			inherent = templaws.inherent
 		if(2)
```
Comment (AnturK) [inline code/datums/ai_laws.dm]: Could just drop it really, it's inferior to weighted pick in every way

Comment (coiax) [inline code/datums/ai_laws.dm]: Backwards compatibility.

**Pull request merged**